Userland/Libraries/LibWeb/Bindings/WindowObject.cpp

```diff
@@ -437,4 +447,114 @@ JS_DEFINE_NATIVE_FUNCTION(WindowObject::inner_height_getter)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return JS::Value(0.0);
+    return JS::Value(static_cast<double>(impl->document().page()->top_level_browsing_context().viewport_scroll_offset().y()));
+}
+
+enum class ScrollBehavior {
+    Auto,
+    Smooth
+};
+
+static void perform_a_scroll(Page& page, double x, double y, ScrollBehavior)
+{
+    // FIXME: Stop any existing smooth-scrolls
+    // FIXME: Implement smooth-scroll
+    page.client().page_did_request_scroll_to({ x, y });
+}
+
+// https://drafts.csswg.org/cssom-view/#dom-window-scroll
+JS_DEFINE_NATIVE_FUNCTION(WindowObject::scroll)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return {};
+    auto& page = *impl->document().page();
+
+    auto viewport_rect = page.top_level_browsing_context().viewport_rect();
+    JS::Value x_value = JS::Value(viewport_rect.x());
+    JS::Value y_value = JS::Value(viewport_rect.y());
+    String behavior_string = "auto";
+
+    if (vm.argument_count() == 1) {
+        auto& options = vm.argument(0).as_object();
+        if (auto left = options.get("left"); !left.is_undefined())
+            x_value = left;
+        if (auto top = options.get("top"); !top.is_undefined())
+            y_value = top;
+        behavior_string = options.get("behavior").to_string_without_side_effects();
+    } else if (vm.argument_count() == 2) {
+        x_value = vm.argument(0);
+        y_value = vm.argument(1);
+    }
+
+    double x = (x_value.is_finite_number() ? x_value.as_double() : 0);
+    double y = (y_value.is_finite_number() ? y_value.as_double() : 0);
+
+    // FIXME: Are we calculating the viewport in the way this function expects?
+    // FIXME: Handle overflow-directions other than top-left to bottom-right
+
+    ScrollBehavior behavior = (behavior_string == "smooth") ? ScrollBehavior::Smooth : ScrollBehavior::Auto;
+    perform_a_scroll(page, x, y, behavior);
+    return {};
+}
+
+// https://drafts.csswg.org/cssom-view/#dom-window-scrollby
+JS_DEFINE_NATIVE_FUNCTION(WindowObject::scroll_by)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return {};
+    auto& page = *impl->document().page();
+
+    JS::Object* options;
+
+    if (vm.argument_count() == 1) {
+        options = &vm.argument(0).as_object();
+    } else if (vm.argument_count() == 2) {
+        options = JS::Object::create(global_object, nullptr);
+        options->set("left", vm.argument(0), ShouldThrowExceptions::No);
+        options->set("top", vm.argument(1), ShouldThrowExceptions::No);
+        options->set("behavior", JS::Value("auto"), ShouldThrowExceptions::No);
+    } else {
+        VERIFY_NOT_REACHED();
+    }
+
+    if (!options->get("left").is_finite_number())
+        options->set("left", JS::Value(0.0), ShouldThrowExceptions::No);
```

Comment:
Missing exception check for `get` and this hides the fact that `left` may not be a number, which should throw/coerce.
```suggestion
    auto left_value = options->get("left");
    if (vm.exception())
        return {};
    auto left = left_value.to_double();
    if (vm.exception())
        return {};
    left = left_value.is_finite_number() ? left : 0.0;
```
Note that you now use the `left` variable from now on, replacing all further use of `options->get("left").as_double()` and `options->set("left")`. This also fixes other issues, as this getter should only be called once and the setter should never be called:
```js
window.scrollBy({ get left() { console.log("left getter called"); return true; }, set left(value) { console.log("left setter called"); throw new Error() }})
```